opt_einsum/paths.py

```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
         path_cost += best[0][1]
 
     return path
+
+
+def ssa_to_linear(ssa_path):
+    """
+    Convert a path with static single assignment ids to a path with recycled
+    linear ids.
+    """
+    ids = np.arange(sum(map(len, ssa_path)), dtype=np.int64)
+    path = []
+    for i, ssa_ids in enumerate(ssa_path):
+        path.append(tuple(ids[ssa_id] for ssa_id in ssa_ids))
+        for ssa_id in ssa_ids:
+            ids[ssa_id:] -= 1
+    return path
+
+
+def linear_to_ssa(path):
+    """
+    Convert a path with recycled linear ids to a path with static single
+    assignment ids.
+    """
+    num_inputs = sum(map(len, path)) - len(path) + 1
+    linear_to_ssa = list(range(num_inputs))
+    new_ids = itertools.count(num_inputs)
+    ssa_path = []
+    for ids in path:
+        ssa_path.append(tuple(linear_to_ssa[id_] for id_ in ids))
+        for id_ in sorted(ids, reverse=True):
+            del linear_to_ssa[id_]
+        linear_to_ssa.append(next(new_ids))
+    return ssa_path
+
+
+def _get_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2):
+    either = k1 | k2
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
```

Comment:
I like this way of keeping track of which indices should be maintained, just a note that it would be good to use the same logic in `optimal` eventually.

Reply:
Yeah `_get_candidate()` is the hotspot in this algorithm, actually this line was the hottest spot. I got about 2x speedup by switching from for loops to `frozenset` logic here.